### PR TITLE
Fix crash with slots and annotated assignments

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

This fixes the issue reported in https://github.com/PyCQA/pylint/issues/5479#issuecomment-988798391

I'm not actually sure if this is the expected output. 
Should this code raise `assigning-non-slot`? If not, the code should be changed. If just added the test code from the issue + a fix to get initial feedback on it.

### CODE REVIEW

**Comment by DanielNoord** [comment]
@mbyrnepr2 Seeing as you opened #5617 and seem to be experienced with `slots`, could I ask you to take a look at this? Does this change look okay? You can look at the CI output to see what the current output for the added test is.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
This need to be rebased following #5497, sorry. You could be able to cherry-pick the commit onto main even if a rebase would conflict..

**Comment by DanielNoord** [comment]
That's an interesting Github bug. `git` saw no issue merging main into this branch. Didn't need to make any changes for this to work..

**Comment by mbyrnepr2** [comment]
Hi @DanielNoord I think your change looks good.

Regarding your question above about whether `assigning-non-slot` should be emitted, I think it also behaves as expected - I.e: `assigning-non-slot` is **not emitted** in your test case.

If I understand this checker, and if it were to emit a warning, then running Python on the example would also result in something similar to: 
`AttributeError: 'Connector' object has no attribute '__reader'`.

As a simplified example of your case, have no issues when running Python:
```
class MyClass:
    __slots__ = ("a", "b")

    @classmethod
    def mymethod(cls):
        cls.c = "x"

MyClass.mymethod()
```

In the following alternative case get an AttributeError:
```
class MyClass2:
    __slots__ = ("a", "b")

    def mymethod(self):
        self.c = "x"

MyClass2().mymethod()
```
`AttributeError: 'MyClass2' object has no attribute 'c'`

Let me know if I've missed anything!

**Comment by DanielNoord** [comment]
@mbyrnepr2 Thanks for your review! Your examples make sense indeed.

I have updated the test options so they should no longer fail on pypy and 3.6. This is ready for review for others!

File: pylint/checkers/classes/class_checker.py [hunk 1]
```diff
@@ -1495,11 +1495,17 @@ def _check_in_slots(self, node):
                     # Properties circumvent the slots mechanism,
                     # so we should not emit a warning for them.
                     return
-                if node.attrname in klass.locals and _has_data_descriptor(
-                    klass, node.attrname
-                ):
-                    # Descriptors circumvent the slots mechanism as well.
-                    return
+                if node.attrname in klass.locals:
+                    for local_name in klass.locals.get(node.attrname, []):
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1499]:
```suggestion
                if node.attrname in klass.locals:
                    for local_name in klass.locals.get(node.attrname):
```
```suggestion
                for local_name in klass.locals.get(node.attrname, []):
```
We know `node.attrname` is in `klass.locals` at this point

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM !

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas  I think that change broke the tests :(

File: pylint/checkers/classes/class_checker.py [hunk 2]
```diff
@@ -1495,11 +1495,17 @@ def _check_in_slots(self, node):
...
+                        statement = local_name.statement(future=True)
+                        if (
+                            isinstance(statement, nodes.AnnAssign)
+                            and not statement.value
+                        ):
+                            return
+                    if _has_data_descriptor(klass, node.attrname):
+                        # Descriptors circumvent the slots mechanism as well.
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1507]:
```suggestion             
                for local_name in klass.locals.get(node.attrname, []):
                    statement = local_name.statement(future=True)
                    if (
                        isinstance(statement, nodes.AnnAssign)
                        and not statement.value
                    ):
                        return
                if node.attrname in klass.locals and _has_data_descriptor(klass, node.attrname):
                    # Descriptors circumvent the slots mechanism as well.
```
Or the first suggestion in the previous comment (no default value)

**Pull request merged**